src/io/der/octet_aligned.rs

```diff
@@ -1,7 +1,161 @@
+use crate::io::per::unaligned::BitRead;
 
-impl DistinguishedRead for OctetBuffer {}
+impl BitRead for OctetBuffer {
+    #[inline]
+    fn read_bit(&mut self) -> Result<bool, Error> {
+        if self.read_position < self.write_position {
+            BitRead::read_bit(&mut (&self.buffer[..], &mut self.read_position))
+        } else {
+            Err(Error::EndOfStream)
+        }
+    }
+
+    #[inline]
+    fn read_bits(&mut self, dst: &mut [u8]) -> Result<(), Error> {
+        BitRead::read_bits(&mut (&self.buffer[..], &mut self.read_position), dst)
+    }
+
+    #[inline]
+    fn read_bits_with_offset(
+        &mut self,
+        dst: &mut [u8],
+        dst_bit_offset: usize,
+    ) -> Result<(), Error> {
+        BitRead::read_bits_with_offset(
+            &mut (&self.buffer[..], &mut self.read_position),
+            dst,
+            dst_bit_offset,
+        )
+    }
+
+    #[inline]
+    fn read_bits_with_len(&mut self, dst: &mut [u8], dst_bit_len: usize) -> Result<(), Error> {
+        BitRead::read_bits_with_len(
+            &mut (&self.buffer[..], &mut self.read_position),
+            dst,
+            dst_bit_len,
+        )
+    }
+
+    #[inline]
+    fn read_bits_with_offset_len(
+        &mut self,
+        dst: &mut [u8],
+        dst_bit_offset: usize,
+        dst_bit_len: usize,
+    ) -> Result<(), Error> {
+        BitRead::read_bits_with_offset_len(
+            &mut (&self.buffer[..], &mut self.read_position),
+            dst,
+            dst_bit_offset,
+            dst_bit_len,
+        )
+    }
+}
+
+#[derive(Debug)]
+pub enum Class {
+    Universal = 0,
+    Application,
+    ContextSpecific,
+    Private
+}
+impl From<u8> for Class {
+    fn from(v: u8) -> Self {
+        match v {
+            x if x == Self::Universal as u8 => Self::Universal,
+            x if x == Self::Application as u8 => Self::Application,
+            x if x == Self::ContextSpecific as u8 => Self::ContextSpecific,
+            x if x == Self::Private as u8 => Self::Private,
+            _ => Self::Universal
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum PC {
+    Primitive = 0,
+    Constructed
+}
+impl From<bool> for PC {
+    fn from(v: bool) -> Self {
+        match v {
+            x if x == false => Self::Primitive,
+            x if x == true => Self::Constructed,
+            _ => Self::Constructed
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Length {
+    Indefinite,
+    Definite(usize),
+    Reserved
+}
+
+impl DistinguishedRead for OctetBuffer {
+
+    fn read_identifier(&mut self) -> Result<(Class, PC, u8), Error> {
+
+        let bit8 = self.read_bit()?;
+        let bit7 = self.read_bit()?;
+        let class = Class::from(((bit8 as u8) << 1) + (bit7 as u8));
+
+        let bit6 = self.read_bit()?;
+        let pc = PC::from(bit6);
+
+        let mut tag_bits = vec![0u8; 1];
+        self.read_bits_with_len(&mut tag_bits[..], 5)?;
+        let tag_number = (tag_bits[0] as u8) >> 3;
```

Comment:
As far as I can tell, DER/BER/CER are designed to read whole octets/bytes and then extract the information from these octets. I imagine something like (totally not tested!):

```rust
let octet = self.read_octet()?;
let bit8 = octet & 0x80 != 0; // required?
let bit7 = octet & 0b40 != 0; // required?
let class = Class:from((octet & 0xC0) >> 2);
```

This would allow us to drop the requirement on `OctetBuffer` for bit-oriented functions and only require simple byte-oriented functions.

Reply:
You are right. I just tried to mimic your code for UPER. Bitwise operation is much better. I'll fix it